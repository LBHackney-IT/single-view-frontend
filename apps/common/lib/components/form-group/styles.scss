@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
