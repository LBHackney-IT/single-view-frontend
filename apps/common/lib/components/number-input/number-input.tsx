import React, { forwardRef, useCallback, useMemo, useState } from "react";

import { Input, InputProps } from "../input";

export interface NumberInputProps extends InputProps {
  min?: number;
  max?: number;
  value?: string | number;
  defaultValue?: string | number;
  maxLength?: number;
  padStart?: number;
}

export const NumberInput = forwardRef<HTMLInputElement, NumberInputProps>(
  function NumberInput(
    {
      onChange,
      onBlur,
      min,
      max,
      value,
      defaultValue,
      maxLength,
      padStart = 0,
      ...props
    },
    ref,
  ) {
    const parser = useCallback(
      (num: string | number) => {
        let numString = String(num).replace(/[^\d]+/g, "");
        if (maxLength !== undefined && maxLength < numString.length) {
          numString = numString.slice(0, maxLength);
        }

        return numString;
      },
      [maxLength],
    );

    const formatter = useCallback(
      (num: string | number) => {
        if (num === "") {
          return "";
        }
        let numInt = parseInt(String(num), 10);
        if (max !== undefined && numInt > max) {
          numInt = max;
        }

        if (min !== undefined && numInt < min) {
          numInt = min;
        }

        return String(numInt).padStart(padStart, "0");
      },
      [min, max, padStart],
    );

    const [output, setOutput] = useState(parser(defaultValue ?? value ?? ""));

    const outputInt = useMemo(() => {
      const target = parseInt(output, 10);
      return !Number.isNaN(target) ? target : undefined;
    }, [output]);

    return (
      <Input
        ref={ref}
        role="spinbutton"
        aria-valuemin={min}
        aria-valuemax={max}
        aria-valuenow={outputInt}
        aria-valuetext={output || undefined}
        value={output}
        onChange={(e) => {
          const update = parser(e.target.value);
          e.target.value = update;
          e.currentTarget.value = update;
          setOutput(update);
          if (onChange) {
            onChange(e);
          }
        }}
        onBlur={(e) => {
          const update = formatter(e.target.value);
          e.target.value = update;
          e.currentTarget.value = update;
          setOutput(update);
          if (onChange) {
            onChange(e);
          }
          if (onBlur) {
            onBlur(e);
          }
        }}
        {...props}
      />
    );
  },
);
