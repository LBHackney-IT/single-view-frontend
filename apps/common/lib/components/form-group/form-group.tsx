import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
