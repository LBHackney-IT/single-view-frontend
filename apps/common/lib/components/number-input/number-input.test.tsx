import React, { useState } from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { FormGroup } from "../form-group";
import { NumberInput } from "./number-input";

const type = (value: string) => {
  const input = screen.getByRole("spinbutton") as HTMLInputElement;
  userEvent.type(input, value);
  return input;
};

test("it passes a11y on single radio", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="NumberInput Test">
      <NumberInput />
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  const { container } = render(<NumberInput />);
  expect(container).toMatchSnapshot();
});

test("it only accepts numbers", () => {
  render(<NumberInput />);
  const input = type("123wa4dv5ggg6");
  expect(input.value).toBe("123456");
});

test("it wont allow characters beyond the maxLength", () => {
  render(<NumberInput maxLength={3} />);
  const input = type("123456");
  expect(input.value).toBe("123");
});

test("it formats min on blur", () => {
  render(<NumberInput min={25} />);
  const input = type("1");
  expect(input.value).toBe("1");
  fireEvent.blur(input);
  expect(input.value).toBe("25");
});

test("it does not format an empty value", () => {
  render(<NumberInput min={25} />);
  const input = screen.getByRole("spinbutton") as HTMLInputElement;
  fireEvent.blur(input);
  expect(input.value).toBe("");
});

test("it formats max on blur", () => {
  render(<NumberInput min={25} max={50} />);
  const input = type("60");
  expect(input.value).toBe("60");
  fireEvent.blur(input);
  expect(input.value).toBe("50");
});

test("it pads the number with 0 on blur", () => {
  render(<NumberInput padStart={4} />);
  const input = type("60");
  expect(input.value).toBe("60");
  fireEvent.blur(input);
  expect(input.value).toBe("0060");
});

test("it can be controlled", () => {
  const Controlled = () => {
    const [value, setValue] = useState("100");
    const [didBlur, setBlur] = useState(false);
    return (
      <>
        <NumberInput
          value={value}
          onChange={(e) => setValue(e.currentTarget.value)}
          onBlur={() => setBlur(!didBlur)}
          max={200}
        />
        <pre data-testid="output">{value}</pre>
        <pre data-testid="blur">{didBlur ? "true" : "false"}</pre>
      </>
    );
  };
  render(<Controlled />);
  const output = screen.getByTestId("output");
  expect(output).toHaveTextContent("100");
  const input = type("3425");
  fireEvent.blur(input);
  const blur = screen.getByTestId("blur");
  expect(blur).toHaveTextContent("true");
});
