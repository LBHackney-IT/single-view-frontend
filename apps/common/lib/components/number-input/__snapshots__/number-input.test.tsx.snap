// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y on single radio 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      NumberInput Test
      
    </label>
    <input
      class="govuk-input lbh-input"
      id="test-field"
      name="test"
      role="spinbutton"
      value=""
    />
  </div>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <input
    class="govuk-input lbh-input"
    role="spinbutton"
    value=""
  />
</div>
`;
